bundles/org.openhab.binding.adorne/src/main/java/org/openhab/binding/adorne/internal/hub/AdorneHubController.java

```diff
@@ -0,0 +1,595 @@
+                this.macAddressCommand = macAddressCommandRet = new CompletableFuture<>();
+            }
+        }
+        return (macAddressCommandRet);
+    }
+
+    private void sendRestCmd(String cmd) {
+        logger.debug("Sending command {}", cmd);
+        synchronized (hubOutLock) {
+            PrintStream hubOut = this.hubOut;
+            if (hubOut != null) {
+                hubOut.print(cmd);
+            } else {
+                throw new IllegalStateException("Can't send command. Adorne Hub connection is not available.");
+            }
+        }
+    }
+
+    /**
+     * Sets a listener to be notified on hub changes.
+     *
+     * @param changeListener instance to be notified
+     */
+    public void setChangeListener(AdorneHubChangeNotify changeListener) {
+        this.changeListener = changeListener;
+    }
+
+    /**
+     * Runs the controller message loop that is interacting with the Adorne Hub by sending commands and listening for
+     * updates
+     */
+    private void msgLoop() {
+        try {
+            JsonObject hubMsg;
+            JsonPrimitive jsonService;
+            String service;
+
+            // Main message loop listening for updates from the hub
+            logger.debug("Starting message loop");
+            while (!shouldStop()) {
+                if (!connect()) {
+                    int sleep = hubReconnectSleep;
+                    logger.debug("Waiting {} seconds before re-attempting to connect.", sleep);
+                    if (hubReconnectSleep < HUB_RECONNECT_SLEEP_MAXIMUM) {
+                        hubReconnectSleep = hubReconnectSleep * 2; // Increase sleep time exponentially
+                    }
+                    restartMsgLoop(sleep);
+                    return;
+                } else {
+                    hubReconnectSleep = HUB_RECONNECT_SLEEP_MINIMUM; // Reset
+                }
+
+                if ((hubMsg = getHubMsg()) == null) {
+                    continue;
+                }
+
+                // Process message based on service type
+                if ((jsonService = hubMsg.getAsJsonPrimitive(HUB_TOKEN_SERVICE)) != null) {
+                    service = jsonService.getAsString();
+                } else {
+                    continue; // Ignore messages that don't have a service specified
+                }
+
+                if (service.equals(HUB_SERVICE_REPORT_ZONE_PROPERTIES)) {
+                    processMsgReportZoneProperties(hubMsg);
+                } else if (service.equals(HUB_SERVICE_ZONE_PROPERTIES_CHANGED)) {
+                    processMsgZonePropertiesChanged(hubMsg);
+                } else if (service.equals(HUB_SERVICE_LIST_ZONE)) {
+                    processMsgListZone(hubMsg);
+                } else if (service.equals(HUB_SERVICE_SYSTEM_INFO)) {
+                    processMsgSystemInfo(hubMsg);
+                }
+            }
+        } catch (RuntimeException e) {
+            logger.error("Hub controller failed", e);
```

Comment:
logging error should be reserved for issues catastrophic enough severely impact the operation of openHab. I don't think a controller error in a binding would qualify as such.
```suggestion
            logger.warn("Hub controller failed", e);
```

Reply:
Done